File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/user/view/model/UserStatsDto.java
package com.cwtsite.cwt.domain.user.view.model;

import com.cwtsite.cwt.domain.user.repository.entity.UserStats;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.util.*;

public class UserStatsDto {

    private Boolean participated;
    private Integer year;
    private Long tournamentId;
    private Integer tournamentMaxRound;
    private Integer round;
    private String locRound;

    public static List<UserStatsDto> toDtos(UserStats userStats) {
        final Iterator<JsonNode> timelineIterator;
        try {
            timelineIterator = new ObjectMapper().readTree("[" + userStats.getTimeline() + "]").elements();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        final ArrayList<UserStatsDto> dtos = new ArrayList<>();
        while (timelineIterator.hasNext()) {
            JsonNode elem = timelineIterator.next();
            final int tournamentMaxRound = elem.get(2).asInt();
            final int round = elem.get(3).asInt();

            final UserStatsDto dto = new UserStatsDto();
            dto.setParticipated(elem.get(3).asInt() != 0);
            dto.setYear(elem.get(1).asInt());
            dto.setTournamentId(elem.get(0).asLong());
            dto.setTournamentMaxRound(tournamentMaxRound);
            dto.setRound(round);
            dto.setLocRound(locRound(tournamentMaxRound, round));
            dtos.add(dto);
        }

        return dtos;
    }

    static String locRound(int tournamentMaxRound, int round) {
        final Map<Integer, String> integerStringHashMap = new HashMap<>();
        int roundOfLastSixteen = tournamentMaxRound - 3;

        integerStringHashMap.put(tournamentMaxRound + 2, "Champion");
        integerStringHashMap.put(tournamentMaxRound + 1, "Runner-up");
        integerStringHashMap.put(tournamentMaxRound, "Third");
        integerStringHashMap.put(tournamentMaxRound - 1, "Semifinal");
        integerStringHashMap.put(tournamentMaxRound - 2, "Quarterfinal");
        integerStringHashMap.put(roundOfLastSixteen, "Last Sixteen");
        integerStringHashMap.put(1, "Group");
        integerStringHashMap.put(0, "Not attended");

        int multiplier = 2;
        for (int i = roundOfLastSixteen - 1; i > 1; i--) {
            integerStringHashMap.put(i, "Last " + Double.valueOf(16 * Math.pow(2, (multiplier - 1))).intValue());
            multiplier++;
        }

        return integerStringHashMap.get(round);
    }

    public Boolean getParticipated() {
        return participated;
    }

    public void setParticipated(Boolean participated) {
        this.participated = participated;
    }

    public Integer getYear() {
        return year;
    }

    public void setYear(Integer year) {
        this.year = year;
    }

    public Long getTournamentId() {
        return tournamentId;
    }

    public void setTournamentId(Long tournamentId) {
        this.tournamentId = tournamentId;
    }

    public Integer getTournamentMaxRound() {
        return tournamentMaxRound;
    }

    public void setTournamentMaxRound(Integer tournamentMaxRound) {
        this.tournamentMaxRound = tournamentMaxRound;
    }

    public Integer getRound() {
        return round;
    }

    public void setRound(Integer round) {
        this.round = round;
    }

    public String getLocRound() {
        return locRound;
    }

    public void setLocRound(String locRound) {
        this.locRound = locRound;
    }
}

File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/user/view/model/UserStatsDto.kt
package com.cwtsite.cwt.domain.user.view.model

import com.cwtsite.cwt.core.toBoolean
import com.cwtsite.cwt.domain.core.DataTransferObject
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper

import java.io.IOException

@DataTransferObject
data class UserStatsDto(
        val participated: Boolean,
        val year: Int,
        val tournamentId: Long,
        val tournamentMaxRound: Int,
        val threeWayFinal: Boolean,
        val round: Int,
        val locRound: String
) {

    companion object {

        fun toDtos(timeline: String): List<UserStatsDto> {
            val timelineIterator: Iterator<JsonNode>
            try {
                timelineIterator = ObjectMapper().readTree("[$timeline]").elements()
            } catch (e: IOException) {
                throw RuntimeException(e)
            }

            val dtos = mutableListOf<UserStatsDto>()

            while (timelineIterator.hasNext()) {
                val elem = timelineIterator.next()
                val threeWayFinal = elem.get(2).asInt().toBoolean()
                val playoffsRoundsMax = elem.get(3).asInt()
                val playoffsRoundAchieved = elem.get(4).asInt()

                dtos.add(UserStatsDto(
                        participated = playoffsRoundsMax != 0,
                        year = elem.get(1).asInt(),
                        tournamentId = elem.get(0).asLong(),
                        tournamentMaxRound = playoffsRoundsMax,
                        round = playoffsRoundAchieved,
                        threeWayFinal = threeWayFinal,
                        locRound = locRound(threeWayFinal, playoffsRoundsMax, playoffsRoundAchieved)
                ))
            }

            return dtos
        }

        private fun locRound(threeWayFinal: Boolean, playoffsRoundMax: Int, achievedRound: Int): String {
            return when (achievedRound) {
                0 -> "Not attended"
                1 -> "Group"
                else -> {
                    if (threeWayFinal) {
                        when (achievedRound) {
                            playoffsRoundMax + 3 -> "Champion"
                            playoffsRoundMax + 2 -> "Runner-up"
                            playoffsRoundMax + 1 -> "Third"
                            playoffsRoundMax -> "Last 6"
                            playoffsRoundMax - 1 -> "Last 12"
                            playoffsRoundMax - 2 -> "Last 24"
                            playoffsRoundMax - 3 -> "Last 48"
                            playoffsRoundMax - 4 -> "Last 96"
                            else -> throw RuntimeException()
                        }
                    } else {
                        when (achievedRound) {
                            playoffsRoundMax + 2 -> "Champion"
                            playoffsRoundMax + 1 -> "Runner-up"
                            playoffsRoundMax -> "Third"
                            playoffsRoundMax - 1 -> "Semifinal"
                            playoffsRoundMax - 2 -> "Quarterfinal"
                            playoffsRoundMax - 3 -> "Last 16"
                            playoffsRoundMax - 4 -> "Last 32"
                            playoffsRoundMax - 5 -> "Last 64"
                            playoffsRoundMax - 6 -> "Last 128"
                            else -> throw RuntimeException()
                        }
                    }
                }
            }
        }
    }
}
